## Question 2

In [2]:
import pandas as pd
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score
from sklearn.model_selection import train_test_split
import numpy as np

## Problem 2

### 2a.

In [5]:
test_df = pd.read_csv("test.csv")
test_df.head()

ID  Gender  Age Occupation  Sleep_Duration  Sleep_Quality  Activity_Level  \
0  330  Female   53   Engineer             8.5              9              30   
1   34    Male   31     Doctor             6.1              6              30   
2   16    Male   29     Doctor             6.0              6              30   
3  326  Female   53   Engineer             8.5              9              30   
4   58    Male   32     Doctor             6.0              6              30   

   Stress_Level BMI_Category Blood_Pressure  Heart_Rate  Daily_Steps  \
0             3       Normal         125/80          65         5000   
1             8       Normal         125/80          72         5000   
2             8       Normal         120/80          70         8000   
3             3       Normal         125/80          65         5000   
4             8       Normal         125/80          72         5000   

  Sleep_Disorder  
0            NaN  
1            NaN  
2            NaN  
3            NaN  
4            NaN

In [6]:
test_df = test_df.drop(columns = ["ID", "Gender", "Occupation", "BMI_Category", "Blood_Pressure", "Sleep_Disorder"])
test_df.head()

Age  Sleep_Duration  Sleep_Quality  Activity_Level  Stress_Level  \
0   53             8.5              9              30             3   
1   31             6.1              6              30             8   
2   29             6.0              6              30             8   
3   53             8.5              9              30             3   
4   32             6.0              6              30             8   

   Heart_Rate  Daily_Steps  
0          65         5000  
1          72         5000  
2          70         8000  
3          65         5000  
4          72         5000

I removed the following columns: ID, Gender, Occupation, BMI_Category, Blood_Pressure, and Sleep_Disorder

### 2b

In [9]:
train_df = pd.read_csv("train.csv")
train_df = train_df.drop(columns = ["ID", "Gender", "Occupation", "BMI_Category", "Blood_Pressure", "Sleep_Disorder"])
train_df.head()

Age  Sleep_Duration  Sleep_Quality  Activity_Level  Stress_Level  \
0   59             8.1              9              75             3   
1   43             6.5              6              45             7   
2   33             6.0              6              30             8   
3   35             7.5              8              60             5   
4   59             8.2              9              75             3   

   Heart_Rate  Daily_Steps  
0          68         7000  
1          72         6000  
2          72         5000  
3          70         8000  
4          68         7000

In [10]:
X_train = train_df.drop(columns = ["Sleep_Quality"])
Y_train = train_df['Sleep_Quality']

model = LinearRegression()
model.fit(X_train, Y_train)

LinearRegression()

In [11]:
intercept = model.intercept_
coefficients = model.coef_
print(f"The models intercept is {intercept} and its coefficients are {coefficients}")

The models intercept is 4.614076015570122 and its coefficients are [ 1.38404261e-02  6.63838357e-01 -5.94043431e-04 -3.22086472e-01
 -2.10175847e-02  9.15066625e-05]


In [12]:
y_pred = model.predict(X_train)

MSE = mean_squared_error(Y_train, y_pred)
r2 = r2_score(Y_train, y_pred)

print(f"The MSE is {MSE} and the r2 score is {r2}")

The MSE is 0.1310026431963264 and the r2 score is 0.9064110852787419


### 2c

In [14]:
X_test = test_df.drop(columns = ["Sleep_Quality"])
Y_test = test_df["Sleep_Quality"]

y_test_pred = model.predict(X_test)

test_MSE = mean_squared_error(Y_test, y_test_pred)
test_r2 = r2_score(Y_test, y_test_pred)

print(f"The MSE is {test_MSE} and the r2 score is {test_r2}")

The MSE is 0.12919025461695402 and the r2 score is 0.9127041171912093


### 2d

In [16]:
train_df2 = pd.read_csv("train.csv")

In [17]:
train_df2 = train_df2.drop(columns = ["ID", "Occupation", "Blood_Pressure", "Sleep_Disorder"])
train_df2

Gender  Age  Sleep_Duration  Sleep_Quality  Activity_Level  Stress_Level  \
0    Female   59             8.1              9              75             3   
1      Male   43             6.5              6              45             7   
2      Male   33             6.0              6              30             8   
3      Male   35             7.5              8              60             5   
4    Female   59             8.2              9              75             3   
..      ...  ...             ...            ...             ...           ...   
295    Male   33             6.1              6              30             8   
296  Female   37             6.1              6              42             6   
297  Female   49             6.1              6              90             8   
298  Female   57             8.2              9              75             3   
299  Female   36             7.2              8              60             4   

      BMI_Category  Heart_Rate  Daily_Steps  
0       Overweight          68         7000  
1       Overweight          72         6000  
2           Normal          72         5000  
3    Normal Weight          70         8000  
4       Overweight          68         7000  
..             ...         ...          ...  
295         Normal          72         5000  
296     Overweight          77         4200  
297     Overweight          75        10000  
298     Overweight          68         7000  
299         Normal          68         7000  

[300 rows x 9 columns]

In [18]:
train_df2["BMI_Category"].unique()

array(['Overweight', 'Normal', 'Normal Weight', 'Obese'], dtype=object)

In [19]:
bmi_map = {
    "Normal": 1,
    "Normal Weight": 1,
    "Overweight": 2,
    "Obese": 3
}

train_df2["BMI_numerical"] = train_df2["BMI_Category"].map(bmi_map)
train_df2 = train_df2.drop(columns = ["BMI_Category"])

In [20]:
gender_map = {
    "Male": 0,
    "Female": 1
}

train_df2['Gender_numerical'] = train_df2['Gender'].map(gender_map)
train_df2 = train_df2.drop(columns = ['Gender'])

train_df2.head()

Age  Sleep_Duration  Sleep_Quality  Activity_Level  Stress_Level  \
0   59             8.1              9              75             3   
1   43             6.5              6              45             7   
2   33             6.0              6              30             8   
3   35             7.5              8              60             5   
4   59             8.2              9              75             3   

   Heart_Rate  Daily_Steps  BMI_numerical  Gender_numerical  
0          68         7000              2                 1  
1          72         6000              2                 0  
2          72         5000              1                 0  
3          70         8000              1                 0  
4          68         7000              2                 1

I chose Gender and BMI Category to encode since both features are relevant to sleep and easy to interpret. Gender was encoded as a binary variable because it only has two categories. BMI Category was encoded using an ordinal scale to reflect increasing body mass, which makes sense for a linear regression model. This approach keeps the encoding simple while still capturing useful information.

In [22]:
X_train2 = train_df2.drop(columns = ['Sleep_Quality'])
Y_train2 = train_df2['Sleep_Quality']

model2 = LinearRegression()

In [23]:
model2.fit(X_train2, Y_train2)
y_train2_pred = model2.predict(X_train2)

train2_MSE = mean_squared_error(Y_train2, y_train2_pred)
train2_r2 = r2_score(Y_train2, y_train2_pred)

In [24]:
test_df2 = pd.read_csv('test.csv')
test_df2 = test_df2.drop(columns = ["ID", "Occupation", "Blood_Pressure", "Sleep_Disorder"])

test_df2["BMI_numerical"] = test_df2["BMI_Category"].map(bmi_map)
test_df2 = test_df2.drop(columns = ["BMI_Category"])
test_df2['Gender_numerical'] = test_df2['Gender'].map(gender_map)
test_df2 = test_df2.drop(columns = ['Gender'])

X_test2 = test_df2.drop(columns = ['Sleep_Quality'])
Y_test2 = test_df2['Sleep_Quality']

In [25]:
y_test2_pred = model2.predict(X_test2)

test2_MSE = mean_squared_error(Y_test2, y_test2_pred)
test2_r2 = r2_score(Y_test2, y_test2_pred)

In [26]:
print(f"For the improved model, the training data had an MSE of {train2_MSE} and an r2 score of {train2_r2}. The testing data had an MSE of {test2_MSE} and an r2 score of {test2_r2}.")

For the improved model, the training data had an MSE of 0.09247555619236834 and an r2 score of 0.9339350204612423. The testing data had an MSE of 0.07243134977099301 and an r2 score of 0.9510570000807061.


The model fits the data very well. In the original model, the training MSE is 0.129 and the R² score is 0.913, meaning the model explains about 91% of the variation in sleep quality and has relatively small prediction error. This shows that linear regression is appropriate for this dataset.

The features that contribute most to the model are sleep-related and health-related variables such as sleep duration, stress level, activity level, and the newly encoded BMI category and gender, since adding these features improves performance.

In the improved model, the training MSE drops to 0.092 and the R2 increases to 0.934, while the testing MSE is 0.072 with an R2 of 0.951. This indicates even better predictive performance and lower error. The testing MSE being slightly lower than the training MSE suggests the model generalizes well and is not overfitting. Overall, the model error is small and the relationship between training and testing MSE shows stable, reliable performance.

## Problem 3

In [29]:
train_df.head()

Age  Sleep_Duration  Sleep_Quality  Activity_Level  Stress_Level  \
0   59             8.1              9              75             3   
1   43             6.5              6              45             7   
2   33             6.0              6              30             8   
3   35             7.5              8              60             5   
4   59             8.2              9              75             3   

   Heart_Rate  Daily_Steps  
0          68         7000  
1          72         6000  
2          72         5000  
3          70         8000  
4          68         7000

In [30]:
X_train3 = train_df.drop(columns=["Sleep_Quality"]).values
Y_train3 = train_df["Sleep_Quality"].values.reshape(-1, 1)

X_train_b3 = np.hstack(
    (np.ones((X_train3.shape[0], 1)), X_train3)
)
X_train_b3

array([[1.0e+00, 5.9e+01, 8.1e+00, ..., 3.0e+00, 6.8e+01, 7.0e+03],
       [1.0e+00, 4.3e+01, 6.5e+00, ..., 7.0e+00, 7.2e+01, 6.0e+03],
       [1.0e+00, 3.3e+01, 6.0e+00, ..., 8.0e+00, 7.2e+01, 5.0e+03],
       ...,
       [1.0e+00, 4.9e+01, 6.1e+00, ..., 8.0e+00, 7.5e+01, 1.0e+04],
       [1.0e+00, 5.7e+01, 8.2e+00, ..., 3.0e+00, 6.8e+01, 7.0e+03],
       [1.0e+00, 3.6e+01, 7.2e+00, ..., 4.0e+00, 6.8e+01, 7.0e+03]])

In [31]:
theta = np.linalg.pinv(X_train_b3.T @ X_train_b3) @ X_train_b3.T @ Y_train3

intercept = theta[0, 0]
coefficients = theta[1:, 0]

print(f"The model's intercept is {intercept} and its coefficients are {coefficients}")

The model's intercept is 4.6140760153907925 and its coefficients are [ 1.38404261e-02  6.63838357e-01 -5.94043432e-04 -3.22086472e-01
 -2.10175847e-02  9.15066625e-05]


In [32]:
y_train_pred3 = X_train_b3 @ theta

In [33]:
def predict_x(x, theta):
    x = np.array(x.reshape(1, -1))
    x_b = np.hstack((np.ones((1, 1)), x))
    y_pred = x_b @ theta
    return float(y_pred[0, 0])

I separated the features and the target variable from the training data. I added a column of ones to the feature matrix so the model could learn an intercept. I then used the normal equation to compute the regression coefficients using matrix multiplication. Lastly, I wrote a function that takes a new data point, adds the intercept term, and multiplies it by the learned coefficients to generate a prediction.

## 3b

In [36]:
y_train_pred_closed = X_train_b3 @ theta

train_mse_closed = mean_squared_error(Y_train3, y_train_pred_closed)
train_r2_closed = r2_score(Y_train3, y_train_pred_closed)

In [37]:
X_test3 = test_df.drop(columns=["Sleep_Quality"]).values
Y_test3 = test_df["Sleep_Quality"].values.reshape(-1, 1)

X_test_b3 = np.c_[np.ones(X_test.shape[0]), X_test3]

y_test_pred_closed = X_test_b3 @ theta

test_mse_closed = mean_squared_error(Y_test3, y_test_pred_closed)
test_r2_closed = r2_score(Y_test3, y_test_pred_closed)

In [38]:
print("Model Comparison")

print("\nClosed-Form Model")
print(f"Train MSE: {train_mse_closed} | R2: {train_r2_closed:}")
print(f"Test MSE: {test_mse_closed} | R2: {test_r2_closed:}")

print("\nScikit-Learn Linear Regression Model")
print(f"Train MSE: {MSE} | R2: {r2}")
print(f"Test MSE: {test_MSE} | R2: {test_r2}")

Model Comparison

Closed-Form Model
Train MSE: 0.13100264319632635 | R2: 0.9064110852787419
Test MSE: 0.1291902546179133 | R2: 0.912704117190561

Scikit-Learn Linear Regression Model
Train MSE: 0.1310026431963264 | R2: 0.9064110852787419
Test MSE: 0.12919025461695402 | R2: 0.9127041171912093


The closed-form implementation produced the same MSE and R² values as the scikit-learn LinearRegression model on both the training and testing sets. This is expected because both methods solve the same least-squares regression problem. The tiny differences in the final decimal places are due to floating-point rounding. Overall, the matching results confirm that my implementation of the normal equation is correct and consistent with the package model.

## Problem 4

### 4a

In [42]:
X_train_4 = train_df[["Sleep_Duration"]].values
Y_train4 = train_df["Sleep_Quality"].values.reshape(-1, 1)

p = 3

X_poly_train4 = np.hstack([X_train_4**k for k in range(1, p+1)])

ones = np.ones((X_poly_train4.shape[0], 1))
X_poly_train_b4 = np.hstack([ones, X_poly_train4])

theta_poly = np.linalg.pinv(X_poly_train_b4.T @ X_poly_train_b4) @ X_poly_train_b4.T @ Y_train4

y_train_pred4 = X_poly_train_b4 @ theta_poly

train_mse4 = mean_squared_error(Y_train4, y_train_pred4)
train_r4 = r2_score(Y_train4, y_train_pred4)

print("Polynomial Regression")
print(f"Degree p = {p}")
print(f"Train MSE: {train_mse4} | Train R2: {train_r4}")

Polynomial Regression
Degree p = 3
Train MSE: 0.2922501114836314 | Train R2: 0.7912151228893163


I implemented polynomial regression by taking the single feature Sleep Duration and expanding it into multiple features using powers of that variable. I then added a column of ones to include the intercept term. After constructing this new feature matrix, I applied the same closed-form least-squares solution from Problem 3 to compute the coefficients. Finally, I generated predictions using matrix multiplication and evaluated the model using MSE and R2.

### 4b

In [45]:
X_test_4 = test_df[["Sleep_Duration"]].values
Y_test4 = test_df["Sleep_Quality"].values.reshape(-1, 1)

results = []

for p in [1, 3, 5]:
    X_poly_train4 = np.hstack([X_train_4**k for k in range(1, p+1)])
    X_poly_test4  = np.hstack([X_test_4**k  for k in range(1, p+1)])

    X_poly_train_b4 = np.hstack([np.ones((X_poly_train4.shape[0], 1)), X_poly_train4])
    X_poly_test_b4  = np.hstack([np.ones((X_poly_test4.shape[0], 1)),  X_poly_test4])

    theta_poly = np.linalg.pinv(X_poly_train_b4.T @ X_poly_train_b4) @ X_poly_train_b4.T @ Y_train4

    y_train_pred4 = X_poly_train_b4 @ theta_poly
    y_test_pred4  = X_poly_test_b4 @ theta_poly

    train_mse4 = mean_squared_error(Y_train4, y_train_pred4)
    train_r4   = r2_score(Y_train4, y_train_pred4)

    test_mse4  = mean_squared_error(Y_test4, y_test_pred4)
    test_r4    = r2_score(Y_test4, y_test_pred4)

    results.append([p, train_mse4, train_r4, test_mse4, test_r4])

print("p | Train MSE | Train R2 | Test MSE | Test R2")
print("--|----------|----------|----------|---------")
for p, tr_mse, tr_r2, te_mse, te_r2 in results:
    print(f"{p}  {round(tr_mse,7)} | {round(tr_r2,7)} | {round(te_mse,7)} | {round(te_r2,7)}")

p | Train MSE | Train R2 | Test MSE | Test R2
--|----------|----------|----------|---------
1  0.3040659 | 0.7827738 | 0.3578027 | 0.7582271
3  0.2922501 | 0.7912151 | 0.3313879 | 0.776076
5  0.2901825 | 0.7926922 | 0.3278314 | 0.7784792


As the polynomial degree p increases from 1 to 5, the training MSE decreases slightly and the training R2 increases, which shows that the model fits the training data better with higher-degree terms. The test MSE also decreases and the test R2 increases across these values of p, meaning the model is also improving on unseen data. This suggests that the relationship between sleep duration and sleep quality is not perfectly linear and that adding polynomial terms up to degree 5 helps capture that pattern without overfitting.

### 4c

As p increases, the training MSE goes down because the model becomes more flexible. The testing MSE also decreases as
p increases, which indicates better generalization rather than overfitting. Overfitting would likely appear only at higher polynomial degrees than the ones tested here.

## Problem 5

## 5a

In [51]:
def addIntercept(X):
    return np.hstack([np.ones((X.shape[0], 1)), X])

def r2Score(Y_true, Y_pred):
    ss_res = np.sum((Y_true - Y_pred) ** 2)
    ss_tot = np.sum((Y_true - np.mean(Y_true)) ** 2)
    return 1 - ss_res / ss_tot

def normalizeTrainTest(X_train, X_test):
    mean = np.mean(X_train, axis=0)
    std = np.std(X_train, axis=0)
    std[std == 0] = 1

    X_train_norm = (X_train - mean) / std
    X_test_norm = (X_test - mean) / std

    return X_train_norm, X_test_norm

def gradientDescent(X_train, Y_train, alpha, num_iters):
    m = X_train.shape[0]
    theta = np.zeros((X_train.shape[1], 1))
    mse_history = []

    for i in range(num_iters):
        Y_pred = X_train @ theta
        gradient = (1/m) * (X_train.T @ (Y_pred - Y_train))
        theta = theta - alpha * gradient
        mse_history.append(mean_squared_error(Y_train, Y_pred))

    return theta, mse_history

In [52]:
Xtrain5 = train_df.drop(columns=["Sleep_Quality"]).values
Ytrain5 = train_df["Sleep_Quality"].values.reshape(-1, 1)

Xtest5 = test_df.drop(columns=["Sleep_Quality"]).values
Ytest5 = test_df["Sleep_Quality"].values.reshape(-1, 1)

XtrainB5 = addIntercept(Xtrain5)

thetaNoNorm5, mseNoNorm5 = gradientDescent(
    XtrainB5, Ytrain5, alpha=0.000001, num_iters=100
)

YtrainPred5 = XtrainB5 @ thetaNoNorm5
print("No normalization")
print("Train MSE:", mean_squared_error(Ytrain5, YtrainPred5))
print("Train R2:", r2Score(Ytrain5, YtrainPred5))

XtrainNorm5, XtestNorm5 = normalizeTrainTest(Xtrain5, Xtest5)

XtrainNormB5 = addIntercept(XtrainNorm5)
XtestNormB5 = addIntercept(XtestNorm5)

thetaNorm5, mseNorm5 = gradientDescent(
    XtrainNormB5, Ytrain5, alpha=0.1, num_iters=100
)

YtrainPredNorm5 = XtrainNormB5 @ thetaNorm5
print("\nWith normalization")
print("Train MSE:", mean_squared_error(Ytrain5, YtrainPredNorm5))
print("Train R2:", r2Score(Ytrain5, YtrainPredNorm5))

No normalization
Train MSE: inf
Train R2: -inf

With normalization
Train MSE: 0.13257587518647596
Train R2: 0.905287160822178


/opt/anaconda3/lib/python3.11/site-packages/numpy/core/_methods.py:118: RuntimeWarning: overflow encountered in reduce
  ret = umr_sum(arr, axis, dtype, out, keepdims, where=where)
/opt/anaconda3/lib/python3.11/site-packages/sklearn/metrics/_regression.py:570: RuntimeWarning: overflow encountered in square
  output_errors = _average((y_true - y_pred) ** 2, axis=0, weights=sample_weight)
/opt/anaconda3/lib/python3.11/site-packages/sklearn/metrics/_regression.py:570: RuntimeWarning: overflow encountered in square
  output_errors = _average((y_true - y_pred) ** 2, axis=0, weights=sample_weight)
/opt/anaconda3/lib/python3.11/site-packages/sklearn/metrics/_regression.py:570: RuntimeWarning: overflow encountered in square
  output_errors = _average((y_true - y_pred) ** 2, axis=0, weights=sample_weight)
/opt/anaconda3/lib/python3.11/site-packages/sklearn/metrics/_regression.py:570: RuntimeWarning: overflow encountered in square
  output_errors = _average((y_true - y_pred) ** 2, axis=0, weight

I implemented gradient descent by initializing the parameter vector to zeros and iteratively updating it using the gradient of the mean squared error loss. I first applied the algorithm without normalizing the features, which caused an error. I then normalized each feature by centering it at the mean and scaling by the standard deviation, which stabilized the updates and allowed the algorithm to converge to a good solution.

## 5b

In [55]:
alphas = [0.01, 0.1, 0.5]
iters = [10, 50, 100]

print("alpha iters | Train MSE | Train R2 | Test MSE | Test R2")
print("------------|----------|----------|----------|---------")

for alpha in alphas:
    for i in iters:
        theta5, _ = gradientDescent(XtrainNormB5, Ytrain5, alpha, i)

        YtrainPred5 = XtrainNormB5 @ theta5
        YtestPred5 = XtestNormB5 @ theta5

        trainMse5 = mean_squared_error(Ytrain5, YtrainPred5)
        trainR25 = r2Score(Ytrain5, YtrainPred5)

        testMse5 = mean_squared_error(Ytest5, YtestPred5)
        testR25 = r2Score(Ytest5, YtestPred5)

        print(
            f"{alpha} {i:>5} | "
            f"{round(trainMse5,7)} | "
            f"{round(trainR25,7)} | "
            f"{round(testMse5,7)} | "
            f"{round(testR25,7)}"
        )

alpha iters | Train MSE | Train R2 | Test MSE | Test R2
------------|----------|----------|----------|---------
0.01    10 | 45.3301379 | -31.3840673 | 42.3987331 | -27.6494895
0.01    50 | 20.1335834 | -13.3835282 | 19.3561324 | -12.0792425
0.01   100 | 7.4323047 | -4.309674 | 7.2712367 | -3.9132888
0.1    10 | 6.7544814 | -3.8254338 | 6.6351712 | -3.4834893
0.1    50 | 0.1361618 | 0.9027254 | 0.1273878 | 0.9139221
0.1   100 | 0.1325759 | 0.9052872 | 0.1276147 | 0.9137687
0.5    10 | 0.1346747 | 0.9037878 | 0.1265983 | 0.9144556
0.5    50 | 0.1311233 | 0.9063249 | 0.1285663 | 0.9131257
0.5   100 | 0.1310044 | 0.9064098 | 0.1291055 | 0.9127614


As the number of iterations increases, the training and testing MSE generally decrease and the R2 increases. Smaller learning rates converge more slowly, while larger learning rates converge faster but can be less stable.

## 5c

The learning rate has a big impact on how fast gradient descent converges. With a small learning rate (α = 0.01), the algorithm converges very slowly, and even after 100 iterations the MSE is still high and the R² values are negative. This shows the model is not fitting the data well yet. With a moderate learning rate (α = 0.1), the error drops much faster. By 50 iterations, the MSE is low and the R2 is close to 0.9 on both the training and testing sets. A larger learning rate (α = 0.5) converges even faster. It achieved similar performance in fewer iterations, but shows small changes as the number of iterations increases. Overall, once the features are normalized and the learning rate is reasonable, gradient descent converges to a solution that is very close to the closed-form model.